src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java

```diff
@@ -337,29 +419,83 @@ void notifyUpdate(V1Service original, V1Service current) {
-                handleApiException(e);
+                handleApiExceptionExceptNotFound(e);
             }
             return job;
         }
 
         @Override
         V1Job applyResource(V1Job original, V1Job current) {
             V1Job job = null;
+            V1Job putJob = getPutObject(original);
             try {
-                job = BATCH_V1_API_INSTANCE.replaceNamespacedJob(getName(), getNamespace(), current, getPretty(), null);
+                job = batchV1ApiInstance.replaceNamespacedJob(
+                        getName(), getNamespace(), putJob, getPretty(), null);
             } catch (ApiException e) {
                 handleApiException(e);
             }
             return job;
         }
 
+        /**
+         * Get PutObject with server-dryRun.
+         * Only some of the fields in the job resource can be modified.
+         * Directly putting the original object of the user will cause some immutable
+         * fields with defaultValue to be modified.
+         * This method is a workaround that populates the default value with the server dryRun implementation.
+         * https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/
+         *
+         * @param original Current Object in Cluster
+         * @return Job With Default Value
+         */
+        V1Job getPutObject(V1Job original) {
+            // Clone Object to avoid modifications to the original object.
+            V1Job dryRunReq = KubernetesJsonUtils.getKubernetesJson().deserialize(
+                    KubernetesJsonUtils.getKubernetesJson().serialize(get()), V1Job.class);
+            // Build dryRun Request Object
+            V1Job dryRunRes = null;
+            dryRunReq.getMetadata().setName(null);
+            dryRunReq.getMetadata().setGenerateName(getName());
+            dryRunReq.getMetadata().setNamespace(getNamespace());
+            try {
+                dryRunRes = batchV1ApiInstance.createNamespacedJob(dryRunReq.getMetadata().getNamespace(),
+                        dryRunReq, null, getPretty(), DRY_RUN_ALL);
+            } catch (ApiException e) {
+                handleApiException(e);
+            }
+            checkNotNull(dryRunRes);
+            // Recover metadata
+            dryRunRes.getMetadata().
+                    name(getName()).
+                    creationTimestamp(original.getMetadata().getCreationTimestamp()).
+                    selfLink(original.getMetadata().getSelfLink()).
+                    uid(original.getMetadata().getUid()).
+                    resourceVersion(original.getMetadata().getUid()).
+                    ownerReferences(original.getMetadata().getOwnerReferences()).
+                    generateName(null);
+            // Recover label/selector
+            String controllerUid = original.getMetadata().getLabels().get(KUBERNETES_CONTROLLER_UID_FIELD);
+            dryRunRes.getMetadata().getLabels().put(KUBERNETES_CONTROLLER_UID_FIELD, controllerUid);
```

Comment:
Any resources to support these implementation? How can we know which properties should we reset here? I am afraid we may miss something here.

Reply:
Have added comments in code